earth_enterprise/src/fusion/tools/boxfilter.cpp

```diff
@@ -104,39 +104,82 @@ const uchar *BoxFilterTiledImage::GetImageTile(int tile_x, int tile_y,
 // Computes a horizontal span of entries in the SAT with a single input
 // value, moving ptrs forward.
-inline void ComputeSATSpanValue(int **psat,   // ptr to first SAT element
-                                int length,   // length of the span
-                                uchar value,  // constant image value (border)
-                                int prev_row_offset) {
+static inline
+void ComputeSATSpanValue(int **psat,   // ptr to first SAT element
+                         int length,   // length of the span
+                         uchar value,  // constant image value (border)
+                         int prev_row_offset) {
   for (int *psat_end = (*psat) + length; (*psat) < psat_end; ++(*psat))
     ComputeSAT((*psat), value, prev_row_offset);
 }
 
 // Computes a horizontal span of entries in the SAT with a row of
 // input values, moving ptrs forward.
-inline void ComputeSATSpanValues(int **psat,  // ptr to first SAT element
-                                 int length,  // length of the span
-                                 const uchar **values,  // ptr to image values
-                                 int prev_row_offset) {
+static inline
+void ComputeSATSpanValues(int **psat,  // ptr to first SAT element
+                          int length,  // length of the span
+                          const uchar **values,  // ptr to image values
+                          int prev_row_offset) {
   for (int *psat_end = (*psat) + length; (*psat) < psat_end;
        ++(*psat), ++(*values))
     ComputeSAT((*psat), **values, prev_row_offset);
 }
 
+// Computes a horizonal span of entries in the SAT that starts from the left
+// edge of the image and covers the width of the box while moving pointers
+// forward. The first box_halfw+1 elements are outside the image and are
+// computed from the border value.  The remaining box_halfw elements are
+// computed from the image.
+static inline
+void ComputeSATSpanFromLeft(int **psat,  // ptr to the row in the SAT
+                             uchar border,  // border vlaue
+                             int prev_row_offset,  // offset to the previous row
+                             int box_halfw,  // (width of box -1) / 2
+                             const uchar **pimg) {  // ptr to image row
+  // Compute the far left value
+  ComputeLeftSAT(psat, border, prev_row_offset);
+  // Compute the padding outside the image
+  ComputeSATSpanValue(psat, box_halfw, border, prev_row_offset);
+  // Compute the remaining half of the box using values from the image
+  ComputeSATSpanValues(psat, box_halfw, pimg, prev_row_offset);
+}
+
 
-// Main function to do the box filtering using a Summed Area Table.
+// Main function to do box filtering.  It is used, for example, when feathering
+// a mask - in that case, each output pixel is the average of all the input
+// pixels within a box centered on the pixel being calculated.
+//
+// This function is heavily optimized.  It uses a Summed Area Table (SAT) to
```

Comment:
can we reference SAT background info  ? may be: https://en.wikipedia.org/wiki/Summed-area_table or some thing similar ?

Reply:
Done.